### PR TITLE
[on-policy distillation] support and related data handling

### PR DESCRIPTION
**Pull request by ahxt**
# PR: **On-Policy Distillation** Support

This PR introduces On-Policy Distillation to the slime framework, extending its reinforcement learning (RL) pipeline to support teacher–student distillation directly within on-policy training.

Thanks to the modular design of slime, integrating On-Policy Distillation is straightforward. In this PR, the teacher model acts as a reward model (RM) by providing teacher log probabilities as the supervision signal. 

### 1. add on_policy_distillation example folder`examples/on_policy_distillation/`
- **`on_policy_distillation.py`** — implements `reward_func` and `post_process_rewards`  
- **`run-qwen3-8B-opd.sh`** — example training script for Qwen3-8B student model and Qwen3-32B as teacher model

### 2. Advantage Estimator Extension (`loss.py`)
- Added `on_policy_distillation` advantage estimator  
- Computes advantages as the difference between teacher and student log probabilities

### 3. Data Pipeline Integration (`rollout.py`, `data.py`)
- Extended rollout data structure to include:
  - `teacher_log_probs`
  - `teacher_token_ids`

### 4. Teacher Model Server
- a separate SGLang server to serve the teacher model and return log probabilities
- only prefill stage for teacher model

### CODE REVIEW

File: examples/on_policy_distillation/run-qwen3-8B-opd.sh [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+####clear after training
+pkill -9 sglang
+sleep 3
+ray stop --force
+pkill -9 ray
+pkill -9 python
+sleep 3
+pkill -9 ray
+pkill -9 python
+
```
Comment (yitianlian) [inline examples/on_policy_distillation/run-qwen3-8B-opd.sh:183]: I think this part could be deleted

File: slime/backends/megatron_utils/data.py [hunk 2]
```diff
@@ -315,7 +315,7 @@ def log_rollout_data(rollout_id: int, args: Namespace, rollout_data: RolloutBatc
         total_lengths = rollout_data["total_lengths"]
 
         for key, val in rollout_data.items():
-            if key == "tokens" or key == "loss_masks" or key == "sample_indices":
+            if key == "tokens" or key == "loss_masks" or key == "sample_indices" or key == "teacher_token_ids" or key == "rewards":
```
Comment (yitianlian) [inline slime/backends/megatron_utils/data.py]: Why we exclude `key=="reward"`

File: slime/backends/megatron_utils/loss.py [hunk 3]
```diff
@@ -286,6 +286,16 @@ def compute_advantages_and_returns(args: Namespace, rollout_data: RolloutBatch)
...
         returns = advantages
 
+    elif args.advantage_estimator == "on_policy_distillation":
+        student_log_probs = log_probs
+        teacher_log_probs = rollout_data.get("teacher_log_probs")
+        response_lengths = rollout_data.get("response_lengths")
+        device = student_log_probs[0].device
+        teacher_log_probs = [t_log_prob.to(device=device) for t_log_prob in teacher_log_probs]
+        teacher_log_probs = [t_log_prob[-response_length:] for t_log_prob, response_length in zip(teacher_log_probs, response_lengths)]
+        advantages = [teacher_log_prob - student_log_prob for teacher_log_prob, student_log_prob in zip(teacher_log_probs, student_log_probs)]
```
Comment (yitianlian) [inline slime/backends/megatron_utils/loss.py]: You should do the slice for teacher logs when creating samples.
Remove teacher token ids.

File: examples/on_policy_distillation/on_policy_distillation.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            resp.raise_for_status()
+            return await resp.json()
+
+
+
+
+def post_process_rewards(args, samples: list[Sample], **kwargs):
+    rewards = [sample.get_reward_value(args) for sample in samples]
+    teacher_log_probs = [torch.tensor([item[0] for item in reward["meta_info"]["input_token_logprobs"][1:]], dtype=torch.float32) for reward in rewards]
+    teacher_token_ids = [torch.tensor([item[1] for item in reward["meta_info"]["input_token_logprobs"][1:]], dtype=torch.int32) for reward in rewards]
```
Comment (yitianlian) [inline examples/on_policy_distillation/on_policy_distillation.py]: Can we do [-response_length:] here?

File: examples/on_policy_distillation/on_policy_distillation.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    for sample, t_log_probs, t_token_ids in zip(samples, teacher_log_probs, teacher_token_ids):
+        sample.teacher_log_probs = t_log_probs
+        sample.teacher_token_ids = t_token_ids
```
Comment (yitianlian) [inline examples/on_policy_distillation/on_policy_distillation.py]: I believe we don’t need teacher token IDs. We should maintain only one token ID list per sample.

File: examples/on_policy_distillation/run-qwen3-8B-opd.sh [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    --port $TEACHER_PORT \
+    --tp 1 \
+    --chunked-prefill-size 4096 \
+    --mem-fraction-static 0.6 \
+    > "$LOG_FILE" 2>&1 &
+
+echo "Starting teacher model server..."
+
+## Wait for the server to be ready
+until curl -sf http://127.0.0.1:$TEACHER_PORT/health_generate > /dev/null; do
```
Comment (yitianlian) [inline examples/on_policy_distillation/run-qwen3-8B-opd.sh]: The IP should use the master address or 0.0.0.0.

File: slime/ray/rollout.py [hunk 7]
```diff
@@ -248,6 +248,12 @@ def _convert_samples_to_train_data(self, samples: Union[list[Sample], list[list[
         if samples[0].train_metadata is not None:
             train_data["metadata"] = [sample.train_metadata for sample in samples]
 
+        if "teacher_log_probs" in samples[0].__dict__:
+            train_data["teacher_log_probs"] = [sample.teacher_log_probs for sample in samples]
+
+        if "teacher_token_ids" in samples[0].__dict__:
+            train_data["teacher_token_ids"] = [sample.teacher_token_ids for sample in samples]
```
Comment (yitianlian) [inline slime/ray/rollout.py]: Remove this

File: slime/backends/megatron_utils/data.py (hunk 2 above)
Comment (ahxt) [reply slime/backends/megatron_utils/data.py]: I’ll add the reward back. However, I’m concerned that the reward here is meaningless—it’s just the average of the teacher log probabilities.

File: slime/backends/megatron_utils/loss.py (hunk 3 above)
Comment (ahxt) [reply slime/backends/megatron_utils/loss.py]: will do. the teacher token ids are for debugging.

File: examples/on_policy_distillation/on_policy_distillation.py (hunk 4 above)
Comment (ahxt) [reply examples/on_policy_distillation/on_policy_distillation.py]: yes. we can and will do.

File: examples/on_policy_distillation/run-qwen3-8B-opd.sh (hunk 6 above)
Comment (ahxt) [reply examples/on_policy_distillation/run-qwen3-8B-opd.sh]: This example only uses one node, so it should be 127.0.0.1 or localhost.

**Comment by ahxt** [comment]
1) I added the first-step loss of this example for reference:

> `(MegatronTrainRayActor pid=46249) step 0: {'train/loss': 0.1438787877559662, 'train/pg_loss': 0.1438787877559662, 'train/entropy_loss': 0.24523311853408813, 'train/pg_clipfrac': 0.0, 'train/ppo_kl': 0.0, 'train/train_rollout_logprob_abs_diff': 0.012880997732281685, 'train/kl_loss': 0.0, 'train/grad_norm': 1.7898532502527087, 'train/lr-pg_0': 1e-06, 'train/lr-pg_1': 1e-06}`


2) I also ran experiments with this code on the OpenThoughts3 dataset. Results (Math500, pass@1, 8 samples):
- Qwen3-8B-Base + SFT: 76%
- Qwen3-8B-Base + SFT + On-Policy-Distillation: 94%

File: examples/on_policy_distillation/run-qwen3-8B-opd.sh (hunk 6 above)
Comment (yitianlian) [reply examples/on_policy_distillation/run-qwen3-8B-opd.sh]: I think it will be better to use 0.0.0.0, as users might directly copy this script and run it on multiple nodes setting.

File: examples/on_policy_distillation/run-qwen3-8B-opd.sh [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+   --rollout-temperature 0.8
+
+   --global-batch-size 64
+   --balance-data
+)
+
+RM_ARGS=(
+   --custom-rm-path examples.on_policy_distillation.on_policy_distillation.reward_func
+   --custom-reward-post-process-path examples.on_policy_distillation.on_policy_distillation.post_process_rewards
+   --rm-url http://127.0.0.1:$TEACHER_PORT/generate
```
Comment (yitianlian) [inline examples/on_policy_distillation/run-qwen3-8B-opd.sh]: Also change here? I think a better script can be:
```bash
teach_IP="0.0.0.0"
teacher_host="13141"
....
```

File: examples/on_policy_distillation/run-qwen3-8B-opd.sh (hunk 6 above)
Comment (ahxt) [reply examples/on_policy_distillation/run-qwen3-8B-opd.sh]: I think 0.0.0.0 won’t work — it should be either the local host (127.0.0.1) or a specific remote host (xxx.xxx.xxx.xx). In this case, the teacher_ip is the local host (127.0.0.1).

I’ll set it like this:
```
teacher_IP="127.0.0.1" # set to your teacher server's IP.
teacher_host="13141"
```

**Review by yitianlian — APPROVED** [review]
LGTM!

**Comment by ahxt** [comment]
Formatted the code to satisfy CI—let me know if anything else is required.”

File: examples/on_policy_distillation/on_policy_distillation.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import aiohttp
+import torch
+
+from slime.utils.types import Sample
+
+
+async def reward_func(args, sample, **kwargs):
+    payload = {
+        "text": sample.prompt + sample.response,
```
Comment (liujiahua123123) [inline examples/on_policy_distillation/on_policy_distillation.py:9]: @ahxt we should probably use `input_ids` here, or there might be discrepancy.
